README.md
# image-renderer
![Operating Systems](https://img.shields.io/badge/OS-Linux%20%7C%20MacOS%20%7C%20Windows-lightgrey)
![License](https://img.shields.io/github/license/teozec/image-renderer)
![Pulls](https://img.shields.io/github/issues-pr/teozec/image-renderer)
![CI](https://img.shields.io/github/workflow/status/teozec/image-renderer/CMake)
![Top Language](https://img.shields.io/github/languages/top/teozec/image-renderer)
![Version](https://img.shields.io/github/v/release/teozec/image-renderer)
![Commits](https://img.shields.io/github/commit-activity/m/teozec/image-renderer)
![Size](https://img.shields.io/github/repo-size/teozec/image-renderer)

image-renderer is a C++ tool to generate photo-realistic images.
It is based on the lectures of the _Numerical tecniques for photorealistic image generation_ course ([EN](https://www.unimi.it/en/education/degree-programme-courses/2021/numerical-tecniques-photorealistic-image-generation), [IT](https://www.unimi.it/it/corsi/insegnamenti-dei-corsi-di-laurea/2021/calcolo-numerico-la-generazione-di-immagini-fotorealistiche)), held by [Prof. Maurizio Tomasi](http://cosmo.fisica.unimi.it/persone/maurizio-tomasi/) in 2021.
A nice presentation of our work is available [here](https://ngrlcu.github.io/Slides/#/)!

## Table of contents

- [Requirements](#requirements)
- [Building](#building)
    - [Standard Unix and Windows](#standard-unix-and-windows)
    - [MacOS (Xcode)](#macOS-(Xcode))
- [Basic tutorial](#basic-tutorial)
	- [A `demo` image](#a-demo-image)
	- [Converting a PFM file to a LDR image with `pfm2ldr`](#Converting-a-PFM-file-to-a-LDR-image-with-pfm2ldr)
	- [`render`ing custom scenes](#rendering-custom-scenes)
	- [Image `stack`ing](#image-stacking)
- [Contributing](#contributing)
- [License](#license)
- [Acknowledgements](#acknowledgements)

## Requirements

The following software is required to build the project:
* [`cmake`](https://cmake.org/) (version >= 3.12)
* A `C++17` compiler.
* [`libgd`](https://libgd.github.io/), compiled with support for:
	* [`libpng`](http://www.libpng.org/pub/png/libpng.html)
	* [`libjpeg`](https://ijg.org/) or [`libjpeg-turbo`](http://www.libpng.org/pub/png/libpng.html)
	* [`libwebp`](https://developers.google.com/speed/webp/)
	* [`libtiff`](http://www.libtiff.org/)
* [`OpenMP`](https://www.openmp.org)

This project also uses the header-only [`Argh!`](https://github.com/adishavit/argh) library, which is already included in this repository in `include/argh.h`.
Therefore, you do not need to install it on your system to compile `image-renderer`.


## Building

After obtaining the source code (e.g. by cloning from [our GitHub repository](https://github.com/teozec/image-renderer)) you can build the project with `cmake`.

### Standard Unix and Windows

Run the following shell commands from the project directory:

```bash
mkdir build
cd build
cmake ..
make
```

The executable files are generated in the `build` directory.

If you use `bash`, you can enable autocompletion for the program running
```bash
source ../tools/bash-completion.bash
```
To make it permanent, you can add it to your `~/.bashrc` file.

### macOS (Xcode)

If you wish to use Xcode on your macOS, you can build the project using:

```bash
mkdir build
cd build
cmake -G "Xcode" ..
```
Now that it is all set you can open up the Xcode project built in the `build` directory and compile it from within the IDE.

If you use `bash`, you can enable autocompletion for the program running (from the `build` directory)
```bash
source ../tools/bash-completion.bash
```
To make it permanent, you can add it to your `~/.bashrc` file.

## Basic tutorial

### A `demo` image
For a quick test, run the following commands from the `build` directory:
```bash
chmod +x ../examples/demo-image.sh
../examples/demo-image.sh
```
and enjoy the `demo.png` file.

![demo](rsc/demo.png)

The script you have just run contains two of the four actions provided by the program: `demo` and `pfm2ldr`.
The first one renders an example scene, and allows you to conifgure it a bit. For example,
```bash
./image-renderer demo --width=500 --height=500 --antialiasing=9 --angleDeg=30 --nMax=5 --roulette=3
```
will render a 500x500 example scene sampling 9 rays per pixel, rotating the camera by 30 degrees on the z axis and considering at most 5 collisions per ray, starting with the Russian roulette algorithm at 3 collisions.
The antialiasing option is useful to sample each pixel properly, avoiding aliasing. Here an example:

![antialiasing](rsc/antialiasing.gif)

Another example of the `demo` action is
```bash
./image-renderer demo -p orthogonal -R flat -o output.pfm
```
It will generate the `output.pfm` image using an orthogonal camera and a flat rendering algorithm.
As you can see, you can use both long and short options: for a list of all supported arguments, you can always use the `--help` flag.

### Converting a PFM file to a LDR image with `pfm2ldr`
The `demo` action, as well as the `render` and `stack` that we will see later, outputs a pfm file. However, we would like to convert it to a LDR image format, such ar `png` or `bmp`.
This is the purpose of the `pfm2ldr` action. A quick usage example:
```bash
./image-renderer pfm2ldr --gamma=1.4 demo.pfm
```
will convert `demo.pfm` to the `demo.png` image, using a correction gamma factor of 1.4.
Of course, you can use a different format if you want: supported formats are `bmp`, `gif`, `jpeg`, `png`, `tiff`, `webp`.
```bash
./image-renderer pfm2ldr --format=webp --outfile=output.webp --luminosity=2.0 --quality=50 input.pfm
```
The previous command will convert `input.pfm` to the `output.webp` image, with a total luminosity of 2.0 and a `webp` compression quality of 50.
In particular, the compression quality is a format-dependent argument (different image formats can be configured differently, so make sure to run `./image-renderer pfm2ldr --help` to get all the details).
The luminosity instead is a tricky parameter: it is what the program should consider as the total brightness of the scene. This means that if a higher value is passed, single pixels will be less bright with respect to the total luminosity, making the scene darker! In general, you should let the renderer calculate and use the real total luminosity; however, in some cases the results may not be optimal and this parameter might be needed.

### `render`ing custom scenes
We have seen how to render a demo image. However, that can soon get boring: this is why we have also provided a `render` action.
To see it at work, you can use
```bash
chmod +x ../examples/cornell.sh
../examples/cornell.sh
```
If you have [GNU Parallel](https://www.gnu.org/software/parallel/) installed, you can also use the faster `../examples/cornell-parallel.sh`.
Depending on your PC, this script may take some hours, but the result is worth the wait:

![cornell](rsc/cornell.png)

The first command in the script makes use of the `render` action, which is used to parse a file describing a scene (aka scenefile) and generate a corresponding `pfm` file.
Many arguments are in common with the `demo` action (except for those about the camera, which is now described in the scenefile). Here is an example:
```bash
./image-renderer render ../examples/scene.txt --float="red:0.3"
```
renders the scene described in `../examples/scene.txt` assigning a custom value to the red variable that appears in the file.
The `--float` option is very useful to render custom animations: if you have [FFmpeg](https://ffmpeg.org/) installed, you can try it with `../examples/animation.sh` (or `../examples/animation-parallel.sh`).

![animation](rsc/animation.gif)

If you wish to learn how to make your own scenefiles, `scene.txt` contains a basic tutorial. Don't forget to `:source ../tools/scenefile.vim` for syntax highlighting, if you use vim.

### Image `stack`ing

Finally, you'd have surely noticed that there is another action in the `cornell.sh` script: `stack`.
It is used to stack many noisy images to get a better looking one by increasing the signal to noise ratio.

![stack](rsc/stack.png)

For example, if you want to stack 10 images saved as `image00.pfm`, `image01.pfm` and so on, you can simply run:
```bash
image-renderer stack image*.pfm
```
This way you will get a `stack.pfm` image in current directory.
You can choose between `mean`and `median` stacking, and also apply sigma-clipping providing a `alpha` factor: for more info please run `image-renderer stack --help`.
This action is very powerful: as a matter of fact it is not only used to get a better signal to noise ratio, but also for rendering blurry images. Here an example:

![blurry](rsc/blurry.png)

## Contributing

If you find any problem or wish to contribute, please open an issue or a pull request on [our GitHub repository](https://github.com/teozec/image-renderer). Thank you!

## License

image-renderer is free software, and it is released under version 3 of the [GNU General Public License](https://www.gnu.org/licenses/gpl-3.0.html).

## Acknowledgements

<img src="https://avatars.githubusercontent.com/u/44500371?v=4" width=100> 
<img src="https://avatars.githubusercontent.com/u/79975678?s=400&u=6770b5f0354ed29bf9a54e7f27a8250bb812c279&v=4" width=100> 

...and many more to come!


CHANGELOG.md
# Version 1.0.0

- Add some example scenefiles and scripts
- Add bash completions ([#PR17](https://github.com/teozec/image-renderer/pull/17)).
- Improve the command line interface ([#PR28](https://github.com/teozec/image-renderer/pull/28)).
- Implement a parser for scene files and a render action ([PR#20](https://github.com/teozec/image-renderer/pull/20), [PR#30](https://github.com/teozec/image-renderer/pull/30), [#29](https://github.com/teozec/image-renderer/issues/29), [#PR32](https://github.com/teozec/image-renderer/pull/32), [#31](https://github.com/teozec/image-renderer/issues/31)).
- Implement `PerlinNoise` and some related textures ([PR#24](https://github.com/teozec/image-renderer/pull/24)).
- Improve `SpecularBRDF` and implement `DielectricBSDF` ([PR#24](https://github.com/teozec/image-renderer/pull/24)).
- Enable multithreading ([PR#25](https://github.com/teozec/image-renderer/pull/25)).
- Bug fix: make Materials work with CSG shapes ([PR#27](https://github.com/teozec/image-renderer/pull/27), [#26](https://github.com/teozec/image-renderer/issues/26)).
- Bug fix: make PathTracer work properly also without Russian roulette ([PR#23](https://github.com/teozec/image-renderer/pull/23), [#21](https://github.com/teozec/image-renderer/issues/21)).
- Implement a DebugRender for better debugging ([PR#22](https://github.com/teozec/image-renderer/pull/22)).

# Version 0.3.0

- Implement a stacking feature, with mean, median and sigma clipping ([PR#15](https://github.com/teozec/image-renderer/pull/15)).
- Add a chair shape
- Implement path-tracing algorithm ([PR#10](https://github.com/teozec/image-renderer/pull/10), [#18](https://github.com/teozec/image-renderer/issues/18), [PR#19](https://github.com/teozec/image-renderer/pull/19)).
- Implement anti-aliasing algorithm ([PR#16](https://github.com/teozec/image-renderer/pull/16)).

# Version 0.2.0

- Implement triangles, boxes and constructive solid geometry ([PR#8](https://github.com/teozec/image-renderer/pull/8)).
- Bug fix: aspect ratio ([PR#14](https://github.com/teozec/image-renderer/pull/14), [#13](https://github.com/teozec/image-renderer/issues/13)).
- Bug fix: shape behaviour ([PR#12](https://github.com/teozec/image-renderer/pull/12), [#11](https://github.com/teozec/image-renderer/issues/11)).
- Implementation of `shape`, containing spheres, planes and world object ([PR#7](https://github.com/teozec/image-renderer/pull/7)).
- Bug fix: image orientation ([PR#6](https://github.com/teozec/image-renderer/pull/6), [#5](https://github.com/teozec/image-renderer/issues/5)).
- Implementation of `camera`, containing cameras, rays and tracer ([PR#4](https://github.com/teozec/image-renderer/pull/4)).
- Implementation of `geometry`, containing geometric objects ([PR#1](https://github.com/teozec/image-renderer/pull/1)).

# Version 0.1.1
- Automated search for libgd in CMakeLists.txt ([PR#3](https://github.com/teozec/image-renderer/pull/3), [#2](https://github.com/teozec/image-renderer/issues/2)).

# Version 0.1.0
- First release of image-renderer.


include/noise.h

/**
THIS IS A DIRECT TRANSLATION TO C++11 FROM THE REFERENCE
JAVA IMPLEMENTATION OF THE IMPROVED PERLIN FUNCTION (see http://mrl.nyu.edu/~perlin/noise/)
THE ORIGINAL JAVA IMPLEMENTATION IS COPYRIGHT 2002 KEN PERLIN

Copyright 2012 Sol from www.solarianprogrammer.com
*/

#ifndef NOISE_H
#define NOISE_H

#include <cmath>
#include <random>
#include <algorithm>
#include <numeric>

//#include "geometry.h"

struct PerlinNoise{
	// The permutation vector
	std::vector<int> p;

public:
	// Initialize with the reference values for the permutation vector
	PerlinNoise() {
		
		// Initialize the permutation vector with the reference values
		p = {
			151,160,137,91,90,15,131,13,201,95,96,53,194,233,7,225,140,36,103,30,69,142,
			8,99,37,240,21,10,23,190, 6,148,247,120,234,75,0,26,197,62,94,252,219,203,117,
			35,11,32,57,177,33,88,237,149,56,87,174,20,125,136,171,168, 68,175,74,165,71,
			134,139,48,27,166,77,146,158,231,83,111,229,122,60,211,133,230,220,105,92,41,
			55,46,245,40,244,102,143,54, 65,25,63,161,1,216,80,73,209,76,132,187,208, 89,
			18,169,200,196,135,130,116,188,159,86,164,100,109,198,173,186, 3,64,52,217,226,
			250,124,123,5,202,38,147,118,126,255,82,85,212,207,206,59,227,47,16,58,17,182,
			189,28,42,223,183,170,213,119,248,152, 2,44,154,163, 70,221,153,101,155,167, 
			43,172,9,129,22,39,253, 19,98,108,110,79,113,224,232,178,185, 112,104,218,246,
			97,228,251,34,242,193,238,210,144,12,191,179,162,241, 81,51,145,235,249,14,239,
			107,49,192,214, 31,181,199,106,157,184, 84,204,176,115,121,50,45,127, 4,150,254,
			138,236,205,93,222,114,67,29,24,72,243,141,128,195,78,66,215,61,156,180 };
		// Duplicate the permutation vector
		p.insert(p.end(), p.begin(), p.end());
	}

	// Generate a new permutation vector based on the value of seed
	PerlinNoise(unsigned int seed) {
		p.resize(256);

		// Fill p with values from 0 to 255
		std::iota(p.begin(), p.end(), 0);

		// Initialize a random engine with seed
		std::default_random_engine engine(seed);

		// Suffle  using the above random engine
		std::shuffle(p.begin(), p.end(), engine);

		// Duplicate the permutation vector
		p.insert(p.end(), p.begin(), p.end());
	}

	float noise(float x, float y, float z) {
		// Find the unit cube that contains the point
		int X = (int) floor(x) & 255;
		int Y = (int) floor(y) & 255;
		int Z = (int) floor(z) & 255;

		// Find relative x, y,z of point in cube
		x -= floor(x);
		y -= floor(y);
		z -= floor(z);

		// Compute fade curves for each of x, y, z
		float u = fade(x);
		float v = fade(y);
		float w = fade(z);

		// Hash coordinates of the 8 cube corners
		int A = p[X] + Y;
		int AA = p[A] + Z;
		int AB = p[A + 1] + Z;
		int B = p[X + 1] + Y;
		int BA = p[B] + Z;
		int BB = p[B + 1] + Z;

		// Add blended results from 8 corners of cube
		float res = lerp(w, lerp(v, lerp(u, grad(p[AA], x, y, z), grad(p[BA], x-1, y, z)), lerp(u, grad(p[AB], x, y-1, z), grad(p[BB], x-1, y-1, z))),	lerp(v, lerp(u, grad(p[AA+1], x, y, z-1), grad(p[BA+1], x-1, y, z-1)), lerp(u, grad(p[AB+1], x, y-1, z-1),	grad(p[BB+1], x-1, y-1, z-1))));
		return (res + 1.0)/2.0;
	}

	float turb(float x, float y, float z, int depth=7) const {
		float accum = 0.f;
		float tempx = x, tempy = y, tempz = z;
		float weight = 1.f;
		PerlinNoise pn;

		for (int i = 0; i < depth; i++) {
			accum += weight*pn.noise(tempx, tempy, tempz);
			weight *= 0.5;
			tempx *= 2;
			tempy *= 2;
			tempz *= 2;
		}
		return fabs(accum);
	}
	
private:

	float fade(float t) { 
		return t * t * t * (t * (t * 6 - 15) + 10);
	}

	float lerp(float t, float a, float b) { 
		return a + t * (b - a); 
	}

	float grad(int hash, float x, float y, float z) {
		int h = hash & 15;
		// Convert lower 4 bits of hash into 12 gradient directions
		float u = h < 8 ? x : y,
			v = h < 4 ? y : h == 12 || h == 14 ? x : z;
		return ((h & 1) == 0 ? u : -u) + ((h & 2) == 0 ? v : -v);
	}
};

#endif //NOISE_H
